### ISSUE TITLE
MemoryMappedInfo cleanup delays Visual Studio shutdown

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#19493]
**tl;dr:** We create instances of `MemoryMappedInfo` (and its nested types). These types are disposable and have finalizers, but in some cases we do not explicitly dispose of instances. The large number of instances created for some user scenarios combined with non-deterministic cleanup resulted in a measurable (~50ms) increase in the time it takes to close Visual Studio for some scenarios.

What follows is a long and detailed breakdown of an investigation that started early Friday morning (5/12) in response to a performance regression detected the day before.

:bulb: When we refer to "meeting internal performance goals" in places like the contributor guide for Roslyn, but don't explain how external contributors are supposed to actually accomplish that, part of what we're talking about is covered by the system explained in detail below. If you were ever interested in what exactly we meant by that, I hope you enjoy the following read. :smile:

### How it started (5/11)

As part of my current role ("grasshopper") on the @dotnet/roslyn-infrastructure team, one of my responsibilities is scheduling and monitoring the incorporation of Roslyn code into Visual Studio itself. The whole process is very new to me, but a reasonably accurate way for a non-Microsoft to view this system is a build farm dedicated to running end-to-end tests prior to taking changes from one large team and applying them to another team. The system as a whole serves as both an early warning system for potential problems and helps identify the team responsible for various types of failures.

One of the toughest types of problems to detect reliably is performance regressions, and anyone who's tried to implement such tests in the past knows what I'm talking about. And the bigger the system under test, the harder it is to produce meaningful results. Small sections of code might be easy to benchmark, but in this case we're talking about ~15+ minute end-to-end user scenario tests of Visual Studio. To keep things flowing smoothly, performance tests in the system allow for some amount of "noise" in the test results. Deviations outside the allowed noise signify major problems where we cannot proceed until they are resolved (I haven't had to deal with one of these so there could be exceptions/special cases that I just don't know about). Deviations "under the degrade bar" do not cause a red flag in the system, still show up on performance reports based on a scoring system. Test scenarios that ran slower than the baseline are given a score based on how close to the bar they are: (actual-baseline)/(bar-baseline), and the highest values show up on the top of the report for manual investigation.

I mentioned "tests" and "scenarios". The two concepts are vital to understanding how we review long and complicated test sequence to identify specific possible regressions. The "test" refers to a complete end-to-end test, which might include opening VS, creating one or more projects, writing some code, building, running, and finally closing the solution and closing Visual Studio itself. Each test is divided into several named scenarios corresponding to specific actions, and target metrics may be defined either for the test as a whole or for specific scenarios.

:bulb: For anyone wanting to set up a system like this someday, I recommend you implement the ability to automatically assign metrics based on what you measure in a shipping release. This "baselining" ability allows you to establish target goals relatively easily that help ensure future releases of your product perform *at least as well* as the previous release, which is a great place to start when discussing development strategies to gradually improve the experience your customers have on a daily basis.

Thursday's test run included one test which did not exceed the bar but seemed "worse than normal", thus starting a manual review. The specific scenario was "Close VS", which in this test occurs after the solution is already closed. The degrade bar is set at 100ms slower than the previous release, and the build was showing a value of 52ms slower. The investigation started with a review of the history of this test using a tool that shows the builds on our team and the builds for the "mainline" that all the teams interact with. This overlay of the two branches is important because it can show with relatively high confidence whether a problem originated on our team (shows up on ours before mainline), or another team (shows up in mainline before ours). Here's the graph going back several builds. The orange branch is our team, and the purple branch is one of the main Visual Studio branches.

![image](https://cloud.githubusercontent.com/assets/1408396/26037111/46c48730-38b2-11e7-8a4e-a22204894d9c.png)

This investigation centered around the apparent new trend where we are sitting roughly 40-50ms slower than our goal line, and which started on our team builds prior to showing up in the main branch.

### Tracking down the problem (5/12)

Tracking down the problem is not always easy, but the two things you can do to be successful on this front are preparation and practice.

#### Preparation

Our end-to-end tests which we use for performance monitoring have PerfView running in the background. I don't have details regarding the exact configuration, but the key items would be:

1. Configure your sampling rate such that you have a "manageable" output data file size over the duration of your test. Since our tests are 15+ minutes, this means a slower sampling rate than the default.
2. Make sure you are enabling items that relate to your user experience. The meaning of this will vary by industry/application, but for me the two items I found valuable were having Thread Time enabled and also having marks inserted into the trace by the end-to-end test driver identifying exactly when a scenario (e.g. Close VS) started and ended.
3. Store the outputs of the PerfView traces for a sufficient amount of time that they can be used for comparative analysis in the future.

#### Practice

I had an advantage in investigating this by knowing three things:

1. The scenario which failed was Close VS, so I could filter the PerfView reports to start and end at the markers placed at the start and end of this scenario. As you can see from the graph, this is a ~300ms window out of the ~15 minute report.
2. Since I was investigating an elapsed time performance failure, the view of most immediate interest was the Thread Time view. I needed to account for the additional 50ms of time, regardless of whether that time was BLOCKED TIME or CPU TIME.
3. Even though this was my first experience ever with this test or scenario, I knew I could most easily track down the time by opening two reports side by side: one from a recent "slow" run, and the other from a recent "fast" run.

### Findings

I started with the Call Tree view examining the main thread. By comparing with the previous run, I pretty quickly narrowed down the additional time on the main thread to the following trace. The highlighted value accounts for all of the new time:

![image](https://cloud.githubusercontent.com/assets/1408396/26037288/d844c974-38b5-11e7-9d57-8c8fcf781a56.png)

As you can see, the main thread was blocked during this time, so to find the root cause we have to find out what it was waiting on. In this case, the answer was pretty easy to infer by looking at the thread activity histograms and asking "what would the CLR be waiting for during `clr!??EEShutdown`.

![image](https://cloud.githubusercontent.com/assets/1408396/26037348/cb182722-38b6-11e7-9a6f-818a9d89a63c.png)

Focusing on the finalizer thread showed two items of particular interest:

![image](https://cloud.githubusercontent.com/assets/1408396/26037410/b1d48458-38b7-11e7-95a0-4834f2403dff.png)

The in the conveniently available CoreCLR sources, we can see that these are the two methods called to [finalize a single object](https://github.com/dotnet/coreclr/blob/76cbc581382ebbcf745dd2b79ac02c419970808c/src/vm/finalizerthread.cpp#L98-L99).

* **clr!WKS::CallFinalizer**: This is slow when you have many finalizers, or when finalizers themselves are slow
* **clr!WKS::LogFinalization**: This has no measurable overhead normally, but a small amount of overhead when ETW events are being recorded. While the call within it, **clr!??ETW::GCLog::SendFinalizeObjectEvent**, probably should have been pruned by our analysis as measurement overhead imposed by PerfView recording, it nevertheless provided valuable insight for me. *Time here means a large number of objects are getting finalized.*

A closer look at **clr!WKS::CallFinalizer** revealed the problem with `MemoryMappedInfo`:

![image](https://cloud.githubusercontent.com/assets/1408396/26037484/e9aaa91a-38b8-11e7-900a-d1cdfdb0d313.png)

In the next step of my investigation, I wanted to determine how many `MemoryMappedInfo` objects were getting finalized when the CLR was trying to shut down. This part was rather difficult, but hopefully in the future it can be improved (filed Microsoft/perfview#200 and Microsoft/perfview#201). I filtered the events view to the start/end time for the scenario, set MaxRet to 110000 (so all events would show), and set the process filter to **devenv**. I then opened the **Microsoft-Windows-DotNETRuntime/GC/FinalizeObject** events, showed the **TypeID** column, and exported the resulting view as a CSV file. After opening the CSV file in Excel, I used a pivot table on the TypeID column to get a histogram which I sorted by the number of times an object of that type ID was finalized. Working from the top, I then examined all available events in PerfView of type **Microsoft-Windows-DotNETRuntime/GC/AllocationTick** to associate type IDs with type names. I only made it through the first 5 items before the sampling rate of AllocationTick stopped showing results.

![image](https://cloud.githubusercontent.com/assets/1408396/26037537/834c1a12-38ba-11e7-9d33-e0bc0d7e0e63.png)

### Follow-up

The solution appears to be two-fold:

1. Reduce the number of memory-mapped file handles that need to be cleaned up. For example, rather than create a memory mapped file for every temporary storage request, we could set a minimum size on the file at 1 MiB, and use smaller views within the file for temporary storage. When a request is too large to fit in the available space at the end of the current file, stop creating storage in that file and move to a new one.
2. Reduce the "weight" of finalization by reducing the size of the object graph referenced by finalizable objects. In particular, avoid making `MemoryMappedInfo` and `SharedReadableStream` themselves finalizable, and instead create a minimal type (or types) derived from `CriticalHandle` which is responsible for releasing the unmanaged resources.

**Comment by sharwell** [comment]
RPS is still showing this as a problem. #19624 provides a more comprehensive long-term solution.
